### PR TITLE
Fix warnings

### LINKED ISSUE #50: Remove #nullable from Program.cs
Once conversion to C# 8 nullable is complete, remove the #nullable declarative from Program.cs

### LINKED ISSUE #206: Test 10.13
Even though powershell returns a 0 exit code, might not be running geocordinates program due to pathing issue on mac/Linux. Need to look in further to actual powershell output.

### LINKED ISSUE #247: Handle Ctrl z to quit
On Tic Tac Toe in chapter 4

### LINKED ISSUE #249: Revisit parse management for null or empty on 5.22 and related listings
lines 22, and other listings like it (`age = int.Parse(ageText);`

### PR DESCRIPTION
**Pull request by BenjaminMichaelis**
Fix all warnings except the 7 expected warnings in the source code. 
Also fixes #206 and fixes #50

Expected warnings:

- Warning CS1030 in 4.58
- Warning CS1030 in 4.61
- Warning CA1507 in 6.20
- Warning CS0108 in 7.11
- Warning CA2013 in 10.04
- Warning CA2013 in 10.04
- Warning CS0612 in 18.25

### CODE REVIEW

File: src/Chapter04/Listing04.44.DoWhileLoop.cs [hunk 1]
```diff
@@ -16,7 +16,8 @@ public static void Main()
                 // Request a move from the current player
                 System.Console.Write(
                     $"\nPlayer {currentPlayer}: Enter move:");
-                string input = System.Console.ReadLine();
+                // TODO: Update listing in Manuscript
+                string input = System.Console.ReadLine() ?? string.Empty;
```
Comment (Keboo) [inline src/Chapter04/Listing04.44.DoWhileLoop.cs]: For this one I think it might be cleaner to simply declare `input` and nullable
```suggestion
                string? input = System.Console.ReadLine();
```

File: src/Chapter04/Listing04.45.ForLoop.cs [hunk 2]
```diff
@@ -11,7 +11,9 @@ public static void Main()
             System.Console.Write("Enter an integer: ");
             // Use long.Parse() to support negative numbers
             // Assumes unchecked assignment to ulong
-            value = (ulong)long.Parse(System.Console.ReadLine());
+            // TODO: Update listing in Manuscript
+            // If ReadLine returns null, use "42" as default input
+            value = (ulong)long.Parse(System.Console.ReadLine() ?? "42");
```
Comment (Keboo) [inline src/Chapter04/Listing04.45.ForLoop.cs]: Since the sample output in the book for this one shows "-42" I would propose we use it for the default.
```suggestion
            value = (ulong)long.Parse(System.Console.ReadLine() ?? "-42");
```

File: src/Chapter04/Listing04.53.ReplacingContinueWithIf.cs [hunk 3]
```diff
@@ -8,7 +8,8 @@ public static void Main()
             bool insideDomain = false;
 
             System.Console.WriteLine("Enter an email address: ");
-            email = System.Console.ReadLine();
+            // TODO: Update listing in Manuscript
```
Comment (Keboo) [inline src/Chapter04/Listing04.53.ReplacingContinueWithIf.cs]: This one doesn't need to be updated as this line is not included in the displayed portion of the listing.

File: src/Chapter04/TicTacToe.cs [hunk 4]
```diff
@@ -59,7 +59,8 @@ public class TicTacToeGame      // Declares the TicTacToeGame class.
             {
                 // Request a move from the current player.
                 System.Console.Write($"\nPlayer {currentPlayer} - Enter move:");
-                input = System.Console.ReadLine();
+                // TODO: Update listing in Manuscript
+                input = System.Console.ReadLine() ?? string.Empty;
```
Comment (Keboo) [inline src/Chapter04/TicTacToe.cs]: For this one I would make `input` nullable, and update the signature of `ValidateAndMove`

File: src/Chapter05/Listing05.02.SimpleMethodCall.cs [hunk 5]
```diff
@@ -10,10 +10,12 @@ public static void Main()
...
 
             System.Console.Write("Enter your first name: ");
-            firstName = System.Console.ReadLine();
+            // TODO: Update listing in Manuscript
+            firstName = System.Console.ReadLine() ?? string.Empty;
 
             System.Console.Write("Enter your last name: ");
-            lastName = System.Console.ReadLine();
+            // TODO: Update listing in Manuscript
+            lastName = System.Console.ReadLine() ?? string.Empty;
```
Comment (Keboo) [inline src/Chapter05/Listing05.02.SimpleMethodCall.cs]: Because the names here are just going into an interpolated string, I would propose just declaring `firstName` and `lastName` as nullable.

File: src/Chapter05/Listing05.12.PassingCommandLineArgumentsToMain.cs [hunk 6]
```diff
@@ -1,5 +1,6 @@
 namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter05.Listing05_12
 {
+    // TODO: Update listing in Manuscript
```
Comment (Keboo) [inline src/Chapter05/Listing05.12.PassingCommandLineArgumentsToMain.cs:3]: I don't think there is an update that is  needed i the manuscript here.

File: src/Chapter05/Listing05.18.CountingLinesGivenADirectory.cs [hunk 7]
```diff
@@ -1,7 +1,6 @@
-#nullable enable
-
 namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter05.Listing05_18
 {
+    // TODO: Update listing in Manuscript
```
Comment (Keboo) [inline src/Chapter05/Listing05.18.CountingLinesGivenADirectory.cs:3]: It doesn't looks like there are any manuscript updates needed here.

File: src/Chapter05/Listing05.22.ConvertingAStringToAnInt.cs [hunk 8]
```diff
@@ -13,10 +13,12 @@ public static void Main()
...
 
             Console.Write("Enter your first name: ");
-            firstName = Console.ReadLine();
+            // TODO: Update listing in Manuscript
+            firstName = Console.ReadLine() ?? string.Empty;
 
             Console.Write("Enter your age: ");
-            ageText = Console.ReadLine();
+            // TODO: Update listing in Manuscript
+            ageText = Console.ReadLine() ?? string.Empty;
```
Comment (Keboo) [inline src/Chapter05/Listing05.22.ConvertingAStringToAnInt.cs]: I would make `firstName` and `ageText` nullable. Also worth noting that both the null and empty string cases will throw on the `int.Parse(ageText)`

File: src/Chapter05/Listing05.23.CatchingAnException.cs [hunk 9]
```diff
@@ -12,10 +12,12 @@ public static int Main(string[] args)
...
 
             Console.Write("Enter your first name: ");
-            firstName = Console.ReadLine();
+            // TODO: Update listing in Manuscript
+            firstName = Console.ReadLine() ?? string.Empty;
 
             Console.Write("Enter your age: ");
-            ageText = Console.ReadLine();
+            // TODO: Update listing in Manuscript
+            ageText = Console.ReadLine() ?? string.Empty;
```
Comment (Keboo) [inline src/Chapter05/Listing05.23.CatchingAnException.cs]: Same here. I would make `firstName` and `ageText` nullable

File: src/Chapter05/Listing05.24.FinallyWithoutCatch.cs [hunk 10]
```diff
@@ -12,10 +12,12 @@ public static int Main()
...
 
             Console.Write("Enter your first name: ");
-            firstName = Console.ReadLine();
+            // TODO: Update listing in Manuscript
+            firstName = Console.ReadLine() ?? string.Empty;
 
             Console.Write("Enter your age: ");
-            ageText = Console.ReadLine();
+            // TODO: Update listing in Manuscript
+            ageText = Console.ReadLine() ?? string.Empty;
```
Comment (Keboo) [inline src/Chapter05/Listing05.24.FinallyWithoutCatch.cs]: Same here. I would make `firstName` and `ageText` nullable

File: src/Chapter05/Listing05.25.GeneralCatchBlocks.cs [hunk 11]
```diff
@@ -12,10 +12,12 @@ public static int Main()
...
 
             System.Console.Write("Enter your first name: ");
-            firstName = System.Console.ReadLine();
+            // TODO: Update listing in Manuscript
+            firstName = System.Console.ReadLine() ?? string.Empty;
 
             System.Console.Write("Enter your age: ");
-            ageText = System.Console.ReadLine();
+            // TODO: Update listing in Manuscript
+            ageText = System.Console.ReadLine() ?? string.Empty;
```
Comment (Keboo) [inline src/Chapter05/Listing05.25.GeneralCatchBlocks.cs]: Same here. I would make `firstName` and `ageText` nullable

File: src/Chapter05/Listing05.28.TryCatch.cs [hunk 12]
```diff
@@ -8,10 +8,12 @@ public static void Main()
...
 
             System.Console.Write("Enter your first name: ");
-            firstName = System.Console.ReadLine();
+            // TODO: Update listing in Manuscript
+            firstName = System.Console.ReadLine() ?? string.Empty;
 
             System.Console.Write("Enter your age: ");
-            ageText = System.Console.ReadLine();
+            // TODO: Update listing in Manuscript
+            ageText = System.Console.ReadLine() ?? string.Empty;
```
Comment (Keboo) [inline src/Chapter05/Listing05.28.TryCatch.cs]: Same here. I would make `firstName` and `ageText` nullable

File: src/Chapter07/Chapter07.csproj [hunk 13]
```diff
@@ -3,16 +3,21 @@
...
+  <ItemGroup>
+    <Compile Remove="Table07.03.TypePatternMatchingWithIsOperator.cs" />
+  </ItemGroup>
   <ItemGroup>
     <Compile Include="..\Shared\Program.cs">
       <Link>Program.cs</Link>
     </Compile>
   </ItemGroup>
+  <ItemGroup>
+    <None Include="Table07.03.TypePatternMatchingWithIsOperator.cs" />
```
Comment (Keboo) [inline src/Chapter07/Chapter07.csproj:19]: I would combine this into the `ItemGroup` above

File: src/Chapter07/Table07.03.TypePatternMatchingWithIsOperator.cs [hunk 14]
```diff
@@ -1,10 +1,45 @@
...
+
+            // ...
+        else (GetObjectById(id) is var result)
+        {
+            // ...
+        }
+
+
+
+    // Was existent in the source code already
```
Comment (Keboo) [inline src/Chapter07/Table07.03.TypePatternMatchingWithIsOperator.cs]:
```suggestion
    // Was existing in the source code already
```

File: src/Chapter10/Listing10.05.OverridingEquals.cs [hunk 15]
```diff
@@ -60,22 +59,20 @@ public bool Equals(Coordinate obj)
         {
             // STEP 1: Check for null if a reference type
             // (e.g., a reference type)
-            // if (obj == null)
+            // if (ReferenceEquals(obj, null)
```
Comment (Keboo) [inline src/Chapter10/Listing10.05.OverridingEquals.cs]: I realize this matches the manuscript, but IMO it would be better if this suggested using pattern matching for the null check to avoid boxing. @MarkMichaelis 
```suggestion
            // if (obj is null)
```

File: src/Chapter10/Listing10.23.RegisteringAFinalizerWithProcessExit.cs [hunk 16]
```diff
@@ -93,12 +94,14 @@ public SampleUnmanagedResource()
             WriteLine("Exiting...");
         }
 
-        #region IDisposable Members
         public void Dispose()
         {
             Dispose(true);
+            
+            // Unregister from the finalization queue.
```
Comment (Keboo) [inline src/Chapter10/Listing10.23.RegisteringAFinalizerWithProcessExit.cs]: I realize this matches the original comment, but I think I would change this wording since the idea of the finalization queue is an implementation detail that is only mostly true (since different runtimes may change this). Also, at this point in the code the object has not been add to it, instead it is preventing itself from getting added.

Perhaps something more like this (stollen from the MS docs):
```suggestion
            // Request that the finalizer not be called for this object.
```

File: src/Chapter13/Listing13.08.UsingMethodNameAsArgument.cs [hunk 17]
```diff
@@ -42,12 +43,18 @@ public static bool GreaterThan(int first, int second)
...
         {
             int i;
+            string? text;
             int[] items = new int[5];
 
             for(i = 0; i < items.Length; i++)
             {
                 Console.Write("Enter an integer: ");
-                items[i] = int.Parse(Console.ReadLine());
+                text = Console.ReadLine();
```
Comment (Keboo) [inline src/Chapter13/Listing13.08.UsingMethodNameAsArgument.cs]: I would combine declaration as assignment of `text` (might even go as far to say the same should be done for `i`)

File: src/Chapter13/Listing13.11.UsingADifferentFuncCompatibleMethod.cs [hunk 18]
```diff
@@ -44,12 +45,18 @@ public static bool GreaterThan(int first, int second)
...
         {
             int i;
+            string? text;
             int[] items = new int[5];
 
             for(i = 0; i < items.Length; i++)
             {
                 Console.Write("Enter an integer: ");
-                items[i] = int.Parse(Console.ReadLine());
+                text = Console.ReadLine();
```
Comment (Keboo) [inline src/Chapter13/Listing13.11.UsingADifferentFuncCompatibleMethod.cs]: I would combine declaration as assignment of `text` (might even go as far to say the same should be done for `i`)

File: src/Chapter13/Listing13.12.CreatingDelegateWithStatementLambda.cs [hunk 19]
```diff
@@ -46,12 +47,18 @@ public static bool GreaterThan(int first, int second)
...
         {
             int i;
+            string? text;
             int[] items = new int[5];
 
             for(i = 0; i < items.Length; i++)
             {
                 Console.Write("Enter an integer: ");
-                items[i] = int.Parse(Console.ReadLine());
+                text = Console.ReadLine();
```
Comment (Keboo) [inline src/Chapter13/Listing13.12.CreatingDelegateWithStatementLambda.cs]: I would combine declaration as assignment of `text` (might even go as far to say the same should be done for `i`)

File: src/Chapter13/Listing13.13.OmittingParameterTypesFromStatementLambdas.cs [hunk 20]
```diff
@@ -8,12 +9,18 @@ public class Program
...
         {
             int i;
+            string? text;
             int[] items = new int[5];
 
             for(i = 0; i < items.Length; i++)
             {
                 Console.Write("Enter an integer:");
-                items[i] = int.Parse(Console.ReadLine());
+                text = Console.ReadLine();
```
Comment (Keboo) [inline src/Chapter13/Listing13.13.OmittingParameterTypesFromStatementLambdas.cs]: I would combine declaration as assignment of `text` (might even go as far to say the same should be done for `i`)

File: src/Chapter13/Listing13.15.ParameterlessStatementLambdas.cs [hunk 21]
```diff
@@ -9,14 +10,15 @@ public static void Main()
...
                 {
-                    string input;
+                    string? input;
                     do
                     {
                         input = Console.ReadLine();
                     }
-                    while(input.Trim().Length == 0);
-                    return input;
+                    while(!string.IsNullOrEmpty(input));
```
Comment (Keboo) [inline src/Chapter13/Listing13.15.ParameterlessStatementLambdas.cs]: The previous code was using `.Trim()` which would remove whitespace. So the equivalent here would be to check against null or whitespace.
```suggestion
                    while(!string.IsNullOrWhiteSpace(input));
```

File: src/Chapter13/Listing13.16.PassingADelegateWithAnExpressionLambda.cs [hunk 22]
```diff
@@ -8,12 +9,18 @@ public class Program
...
         {
             int i;
+            string? text;
             int[] items = new int[5];
 
             for(i = 0; i < items.Length; i++)
             {
                 Console.Write("Enter an integer:");
-                items[i] = int.Parse(Console.ReadLine());
+                text = Console.ReadLine();
```
Comment (Keboo) [inline src/Chapter13/Listing13.16.PassingADelegateWithAnExpressionLambda.cs]: I would combine declaration as assignment of `text` (might even go as far to say the same should be done for `i`)

File: src/Chapter13/Listing13.17.PassingAnAnonymousMethodInCSharpTwo.cs [hunk 23]
```diff
@@ -8,12 +9,18 @@ public class Program
...
         {
             int i;
+            string? text;
             int[] items = new int[5];
 
             for(i = 0; i < items.Length; i++)
             {
                 Console.Write("Enter an integer:");
-                items[i] = int.Parse(Console.ReadLine());
+                text = Console.ReadLine();
```
Comment (Keboo) [inline src/Chapter13/Listing13.17.PassingAnAnonymousMethodInCSharpTwo.cs]: I would combine declaration as assignment of `text` (might even go as far to say the same should be done for `i`)

File: src/Chapter13/Listing13.18.UsingVarianceForDelegates.cs [hunk 24]
```diff
@@ -17,9 +17,10 @@ public static void Main()
 
             Action<string> narrowAction = broadAction;
 
+            // TODO: Update listing in Manuscript
             // Covariance
             Func<string> narrowFunction =
-                () => Console.ReadLine();
+                () => 8654091021.ToString();
```
Comment (Keboo) [inline src/Chapter13/Listing13.18.UsingVarianceForDelegates.cs]: If we are just going to return a hard coded string I would just use a string literal. 
```suggestion
                () => "8654091021";
```

File: src/Chapter13/Listing13.19.CSharpEquivalentOfCILGeneratedByTheCompilerForLambdaExpressions.cs [hunk 25]
```diff
@@ -32,12 +33,18 @@ public class DelegateSample
...
         {
             int i;
+            string? text;
             int[] items = new int[5];
 
             for(i = 0; i < items.Length; i++)
             {
                 Console.Write("Enter an integer:");
-                items[i] = int.Parse(Console.ReadLine());
+                text = Console.ReadLine();
```
Comment (Keboo) [inline src/Chapter13/Listing13.19.CSharpEquivalentOfCILGeneratedByTheCompilerForLambdaExpressions.cs]: I would combine declaration as assignment of `text` (might even go as far to say the same should be done for `i`)

File: src/Chapter13/Listing13.20.UsingAnOuterVariableInALambdaExpression.cs [hunk 26]
```diff
@@ -8,13 +9,19 @@ public class Program
...
             int i;
+            string? text;
             int[] items = new int[5];
             int comparisonCount = 0;
 
             for(i = 0; i < items.Length; i++)
             {
                 Console.Write("Enter an integer:");
-                items[i] = int.Parse(Console.ReadLine());
+                text = Console.ReadLine();
```
Comment (Keboo) [inline src/Chapter13/Listing13.20.UsingAnOuterVariableInALambdaExpression.cs]: I would combine declaration as assignment of `text` (might even go as far to say the same should be done for `i`)

File: src/Chapter13/Listing13.21.CSharpEquivalentOfCILCodeGeneratedByCompilerForOuterVariables.cs [hunk 27]
```diff
@@ -24,11 +25,17 @@ public static void Main()
...
             locals.comparisonCount = 0;
             int[] items = new int[5];
+            string? text;
 
-            for(i = 0; i < items.Length; i++)
+            for (i = 0; i < items.Length; i++)
             {
                 Console.Write("Enter an integer:");
-                items[i] = int.Parse(Console.ReadLine());
+                text = Console.ReadLine();
```
Comment (Keboo) [inline src/Chapter13/Listing13.21.CSharpEquivalentOfCILCodeGeneratedByCompilerForOuterVariables.cs]: I would combine declaration as assignment of `text` (might even go as far to say the same should be done for `i`)

File: src/Chapter14/Listing14.03.ConnectingThePublisherAndSubscribers.cs [hunk 28]
```diff
@@ -11,7 +12,7 @@ public static void Main()
             Thermostat thermostat = new Thermostat();
             Heater heater = new Heater(60);
             Cooler cooler = new Cooler(80);
-            string temperature;
+            string? temperature;
```
Comment (Keboo) [inline src/Chapter14/Listing14.03.ConnectingThePublisherAndSubscribers.cs]: I would combine declaration as assignment of `temperature`

File: src/Chapter14/Listing14.09.OnTemperatureChangedThrowingAnException.cs [hunk 29]
```diff
@@ -11,7 +12,7 @@ public static void Main()
             Thermostat thermostat = new Thermostat();
             Heater heater = new Heater(60);
             Cooler cooler = new Cooler(80);
-            string temperature;
+            string? temperature;
```
Comment (Keboo) [inline src/Chapter14/Listing14.09.OnTemperatureChangedThrowingAnException.cs]: I would combine declaration as assignment of `temperature`

File: src/Chapter14/Listing14.10.HandlingExceptionsFromSubscribers.cs [hunk 30]
```diff
@@ -59,7 +60,7 @@ public static void Main()
                 Thermostat thermostat = new Thermostat();
                 Heater heater = new Heater(60);
                 Cooler cooler = new Cooler(80);
-                string temperature;
+                string? temperature;
```
Comment (Keboo) [inline src/Chapter14/Listing14.10.HandlingExceptionsFromSubscribers.cs]: I would combine declaration as assignment of `temperature`

File: src/Chapter14/Listing14.11.UsingTheAssignmentOperationRatherThanPlusEquals.cs [hunk 31]
```diff
@@ -12,7 +13,7 @@ public static void Main()
             Thermostat thermostat = new Thermostat();
             Heater heater = new Heater(60);
             Cooler cooler = new Cooler(80);
-            string temperature;
+            string? temperature;
```
Comment (Keboo) [inline src/Chapter14/Listing14.11.UsingTheAssignmentOperationRatherThanPlusEquals.cs]: I would combine declaration as assignment of `temperature`

File: src/Chapter16/Listing16.05.DeferredExecutionAndQueryExpressions2.cs [hunk 32]
```diff
@@ -55,7 +55,8 @@ where keyword.Contains('*')
 
             // Retrieve the count from the cached collection
             Console.WriteLine(
-                $"7. selectionCache count={ selectionCache.Count() }");
+                // TODO: Update listing in Manuscript
```
Comment (Keboo) [inline src/Chapter16/Listing16.05.DeferredExecutionAndQueryExpressions2.cs:58]: I don't think there is a manuscript change needed here.

File: src/Chapter18/18.11.SpecifyingAReturnAttribute.DescriptionAttribute.cs [hunk 33]
```diff
@@ -4,8 +4,8 @@ namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter18.Listing18_11
 {
     public class DescriptionAttribute : Attribute
     {
-        private string description;
-		public string Description { get { return Description; } }
+        private readonly string description;
+		public string Description { get { return description; } }
```
Comment (Keboo) [inline src/Chapter18/18.11.SpecifyingAReturnAttribute.DescriptionAttribute.cs]: I realize this matches the original, but we should fix the naming of the field.
```suggestion
        private readonly string _Description;
	public string Description { get { return _Description; } }
```

File: src/Chapter19/Listing19.07.RegisteringForUnhandledExceptions.cs [hunk 34]
```diff
@@ -6,12 +6,12 @@
 
     public class Program
     {
-        public static Stopwatch _Clock = new Stopwatch();
+        private static Stopwatch _Clock { get;  } = new Stopwatch();
```
Comment (Keboo) [inline src/Chapter19/Listing19.07.RegisteringForUnhandledExceptions.cs]: We should fix the name of the property here. 
```suggestion
        private static Stopwatch Clock { get; } = new Stopwatch();
```

File: src/Chapter20/Listing20.12.CustomMethodWithProgressSupport.cs [hunk 35]
```diff
@@ -10,10 +11,10 @@ class Program
         static public Task<Process> RunProcessAsync(
             string fileName,
             string arguments = "",
+            IProgress<ProcessProgressEventArgs>? progress = null,
+                object? objectState = null,
```
Comment (Keboo) [inline src/Chapter20/Listing20.12.CustomMethodWithProgressSupport.cs]: nit: whitespace
```suggestion
            object? objectState = null,
```

File: src/Chapter23/Listing23.20.DesignatingABlockForUnsafeCode.cs [hunk 36]
```diff
@@ -54,7 +54,7 @@ public unsafe static int Main()
         }
     }
 
-    public class VirtualMemoryPtr : System.Runtime.InteropServices.SafeHandle
+    public class VirtualMemoryPtr : System.Runtime.InteropServices.SafeHandle, IDisposable
```
Comment (Keboo) [inline src/Chapter23/Listing23.20.DesignatingABlockForUnsafeCode.cs]: The interface declaration is technically not needed here as the base class already implements `IDisposable`. This also means you don't need to create the full dispose pattern in this class.

File: src/Chapter23/Listing23.20.DesignatingABlockForUnsafeCode.cs [hunk 37]
```diff
@@ -90,15 +90,37 @@ public override bool IsInvalid
...
+                    VirtualMemoryManager.VirtualFreeEx(ProcessHandle,
+                        AllocatedPointer, MemorySize);
+                }
                 Disposed = true;
-                GC.SuppressFinalize(this);
-                VirtualMemoryManager.VirtualFreeEx(ProcessHandle,
-                    AllocatedPointer, MemorySize);
             }
-            return true;
+        }
```
Comment (Keboo) [inline src/Chapter23/Listing23.20.DesignatingABlockForUnsafeCode.cs]: This class is deriving from SafeHandle, so it is a bit of a special case in terms of memory clean up. Because the base class implements the IDisposable pattern, it exposes the regular Dispose(bool) method that you can override to clean up. In addition, the abstract base class exposes the ReleaseHandle method (which is also designed for cleanup). Making things just a little bit harder, if you dig into the base dispose method you will find that it ends in some extern calls that get implemented by the runtime (making it tricky to tell what is actually happening). Lot of details in these docs here: SafeHandle Class (System.Runtime.InteropServices) | Microsoft Docs. Take note of the section at the very bottom titled:  Notes to Implementers 
What is convenient here is we can assume that MS has implemented the dispose correctly, so the amount of code we need to write is significantly reduced. Let’s start by looking at the regular Dispose(bool) (I am just using the default VS generates). This will be functionally similar to what SafeHandle will do:
```C#
protected virtual void Dispose(bool disposing)
{
    if (!_disposedValue)
    {
        if (disposing)
        {
            // TODO: dispose managed state (managed objects)
        }

        // TODO: free unmanaged resources (unmanaged objects) and override finalizer
        // TODO: set large fields to null
        _disposedValue = true;
    }
}
```
The Dispose implementation inside of SafeHandle will ensure that Dispose gets appropriately called (passing the right disposing value). In the TODO section for freeing unmanaged resources is where it will call ReleaseHandle (expecting that it will do the needful to cleanup unmanaged handles). You will also notice that it does this regardless of whether the disposing flag. So for our code we only need to implement ReleaseHandle, and don’t need to do any of the regular disposable members as those are already handled for us.
So for our code here the implementation we want should look like this:
```C#
protected override bool ReleaseHandle()
{
    Disposed = true;
    VirtualMemoryManager.VirtualFreeEx(ProcessHandle, AllocatedPointer, MemorySize);
}
```

File: src/Shared/Program.cs [hunk 38]
```diff
@@ -84,24 +84,24 @@ public static void Main(string[] args)
...
-                        output = InvokeMethodUsingReflection(method,arguments).GetAwaiter().GetResult();
+                        output = await InvokeMethodUsingReflectionAsync(method, arguments);
                     });
                     task.Wait();
                 }
                 else
                 {
                     // TODO: Change to use async/await.
-                    output = InvokeMethodUsingReflection(method, arguments).GetAwaiter().GetResult();
+                    output = await InvokeMethodUsingReflectionAsync(method, arguments);
```
Comment (Keboo) [inline src/Shared/Program.cs]: I think entire if-condition can be replaced with:
```C#
output = await InvokeMethodUsingReflectionAsync(method, arguments);
```

The extra Tasks don't appear to be adding anything, and the STAAttribute is getting ignored and wont be used anyway. If all of the tests are passing we should b esafe to remove it.

File: src/Chapter23/Listing23.20.DesignatingABlockForUnsafeCode.cs (hunk 36 above)
Comment (BenjaminMichaelis) [reply src/Chapter23/Listing23.20.DesignatingABlockForUnsafeCode.cs]: I got build errors by removing this because lower down we have this function within here that calls IDisposable so I am keeping it for now:
```
        void IDisposable.Dispose()
        {
            // Do not change this code. Put cleanup code in 'Dispose(bool disposing)' method
            Dispose(disposing: true);
            GC.SuppressFinalize(this);
        }
```

**Comment by BenjaminMichaelis** [comment]
Just wanted to get a double check over these last like 4 commits since the review just as a final double check

File: src/Chapter23/Listing23.20.DesignatingABlockForUnsafeCode.cs (hunk 36 above)
Comment (Keboo) [reply src/Chapter23/Listing23.20.DesignatingABlockForUnsafeCode.cs]: Correct, I would suggest that we should remove that as well

File: src/Chapter20.Tests/Listing20.11.ImplementingCustomAsynchronousMethod.Tests.cs [hunk 39]
```diff
@@ -1,10 +1,8 @@
 using Microsoft.VisualStudio.TestTools.UnitTesting;
 using System;
 using System.Text.RegularExpressions;
-using System.Threading.Tasks;
-using AddisonWesley.Michaelis.EssentialCSharp.Chapter20.Listing20_09;
 
-namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter20.Listing20_11.Tests
+namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter20.Listing20_09.Tests
```
Comment (Keboo) [inline src/Chapter20.Tests/Listing20.11.ImplementingCustomAsynchronousMethod.Tests.cs:5]: Was this change intentional? Is there a reason this doesn't match the file name?

File: src/Chapter23/Listing23.20.DesignatingABlockForUnsafeCode.cs [hunk 40]
```diff
@@ -91,14 +91,20 @@ public override bool IsInvalid
...
+        {
+            // Do not change this code. Put cleanup code in 'Dispose(bool disposing)' method
+            Dispose(disposing: false);
+        }
+
+        void IDisposable.Dispose()
+        {
+            // Do not change this code. Put cleanup code in 'Dispose(bool disposing)' method
+            Dispose(disposing: true);
+            GC.SuppressFinalize(this);
```
Comment (Keboo) [inline src/Chapter23/Listing23.20.DesignatingABlockForUnsafeCode.cs]: I would propose removing these as well since the SafeHandle already defines them.

File: src/Shared/Program.cs [hunk 41]
```diff
@@ -84,24 +84,10 @@ public static void Main(string[] args)
...
-                    task.Wait();
-                }
-                else
-                {
-                    // TODO: Change to use async/await.
-                    output = InvokeMethodUsingReflection(method, arguments).GetAwaiter().GetResult();
-                }
-                if(output is { })
+
+                output = await InvokeMethodUsingReflectionAsync(method, arguments);
```
Comment (Keboo) [inline src/Shared/Program.cs]: nit: Combine declaration and assignment.

```C#
string? output = await InvokeMethodUsingReflectionAsync(method, arguments);
```

**Review by Keboo — COMMENTED** [review]
Just a couple more comments then it is looking good.

File: src/Chapter20.Tests/Listing20.11.ImplementingCustomAsynchronousMethod.Tests.cs (hunk 39 above)
Comment (BenjaminMichaelis) [reply src/Chapter20.Tests/Listing20.11.ImplementingCustomAsynchronousMethod.Tests.cs:5]: The file name was going to get updated here: #246 but I can do it in this one instead just for continuity

Comment (Keboo) [reply src/Chapter20.Tests/Listing20.11.ImplementingCustomAsynchronousMethod.Tests.cs:5]: Ah, either is fine by me.

File: src/Chapter05/Listing05.12.PassingCommandLineArgumentsToMain.cs (hunk 6 above)
Comment (BenjaminMichaelis) [reply src/Chapter05/Listing05.12.PassingCommandLineArgumentsToMain.cs:3]: This one does need to be for lines 4-6

File: src/Chapter05/Listing05.18.CountingLinesGivenADirectory.cs (hunk 7 above)
Comment (BenjaminMichaelis) [reply src/Chapter05/Listing05.18.CountingLinesGivenADirectory.cs:3]: Line 1 in the book needs to be removed.

File: src/Chapter05/Listing05.22.ConvertingAStringToAnInt.cs (hunk 8 above)
Comment (BenjaminMichaelis) [reply src/Chapter05/Listing05.22.ConvertingAStringToAnInt.cs]: Made an issue to follow up the parse

File: src/Chapter10/Listing10.23.RegisteringAFinalizerWithProcessExit.cs (hunk 16 above)
Comment (BenjaminMichaelis) [reply src/Chapter10/Listing10.23.RegisteringAFinalizerWithProcessExit.cs]: Sounds great to me!

File: src/Chapter16/Listing16.05.DeferredExecutionAndQueryExpressions2.cs (hunk 32 above)
Comment (BenjaminMichaelis) [reply src/Chapter16/Listing16.05.DeferredExecutionAndQueryExpressions2.cs:58]: This needs to be switched from a function to a property in the book

File: src/Chapter10/Listing10.05.OverridingEquals.cs (hunk 15 above)
Comment (BenjaminMichaelis) [reply src/Chapter10/Listing10.05.OverridingEquals.cs]: Needs to be done throughout the book, but changed it here

**Review by Keboo — APPROVED** [review]

**Comment by BenjaminMichaelis** [comment]
@Keboo Should have them all resolved now
